File: 프로그래머스/2/388353. 지게차와 크레인/지게차와 크레인.js
    
const dx = [-1,1,0,0]
const dy = [0,0,-1,1]

function solution(storage, requests) {
    const n = storage.length
    const m = storage[0].length
    
    // Padding
    const newStorage = [Array(m + 2).fill('0')]
    for (const s of storage){
        const row = ['0']
        for (let i = 0; i < m; i ++){
            row.push(s[i])
        }
        row.push('0')
        newStorage.push(row)
    }
    newStorage.push(Array(m + 2).fill('0'))
    
    // Request Handling
    requests.forEach(r => {
        if (r.length === 1){
            fork(r)
        }
        else {
            crain(r)
        }
    })
    
    function update () {
        const queue = []
        const visited = Array.from({length: n + 2},() => Array(m + 2).fill(false))
        queue.push([0,0])
        visited[0][0] = true
        
        while (queue.length > 0){
            const [x,y] = queue.shift()
            
            for (let d = 0; d < 4; d ++){
                const nx = x + dx[d]
                const ny = y + dy[d]
                
                if (0 <= nx && nx < n + 2 && 0 <= ny && ny < m + 2 && !visited[nx][ny]){
                    if (newStorage[nx][ny] === '0'){
                        visited[nx][ny] = true
                        queue.push([nx,ny])
                    }
                    else if (newStorage[nx][ny] == '1'){
                        newStorage[nx][ny] = '0'
                        visited[nx][ny] = true
                        queue.push([nx,ny])
                    }
                }
            }
        }
        return
    }
    
    function fork (request) {
        const buffer = []
        
        for (let i = 0; i < n + 2; i ++){
            for (let j = 0; j < m + 2; j ++){
                if (newStorage[i][j] === request){
                    for (let d = 0; d < 4; d ++){
                        const nx = i + dx[d]
                        const ny = j + dy[d]
                        // 주위에 빈칸이 있는경우
                        if (0 <= nx && nx < n + 2 && 0 <= ny && ny < m + 2 && newStorage[nx][ny] === '0'){
                            buffer.push([i,j])
                            break
                        }
                    }
                }
            }
        }
        
        for (const [x,y] of buffer){
            newStorage[x][y] = '0'
        }
        update()
        return
    }
    
    function crain (request) {
        const target = request[0]
        const buffer = []
        
        for (let i = 0; i < n + 2; i ++){
            for (let j = 0; j < m + 2; j ++){
                if (newStorage[i][j] === target){
                    buffer.push([i,j])
                }
            }
        }
        
        for (const [x,y] of buffer){
            newStorage[x][y] = '1'
        }
        
        update()
        return
    }
    
    for (const s of newStorage) console.log(...s)
    
    let answer = 0
    for (let i = 1; i < n + 1; i ++){
        for (let j = 1; j < m + 1; j ++){
            if (newStorage[i][j] !== '0' && newStorage[i][j] !== '1') answer++
        }
    }
    return answer
}

File: 프로그래머스/2/388353. 지게차와 크레인/지게차와 크레인.py
from collections import deque
dx = [-1,1,0,0]
dy = [0,0,-1,1]

def solution(storage, requests):
    n = len(storage)
    m = len(storage[0])
    new_storage = [['0'] * (m + 2) for _ in range(n + 2)]
    
    for i in range(1, n + 1):
        for j in range(1, m + 1):
            new_storage[i][j] = storage[i - 1][j - 1]

    def update():
        nonlocal new_storage
        queue = deque()
        queue.append((0,0))
        visited = [[False] * (m + 2) for _ in range(n + 2)]
        visited[0][0] = True
        
        while queue:
            x,y = queue.popleft()
            
            for d in range(4):
                nx,ny = x + dx[d], y + dy[d]
                if 0 <= nx < n + 2 and 0 <= ny < m + 2 and not visited[nx][ny]:
                    if new_storage[nx][ny] == '0':
                        visited[nx][ny] = True
                        queue.append((nx,ny))
                    elif new_storage[nx][ny] == '1':
                        visited[nx][ny] = True
                        queue.append((nx,ny))
                        new_storage[nx][ny] = '0'
            
    def fork(request):
        buffer = []
        nonlocal new_storage
        for i in range(n + 2):
            for j in range(m + 2):
                if new_storage[i][j] == request:
                    for d in range(4):
                        nx,ny = i + dx[d], j + dy[d]
                        if 0 <= nx < n + 2 and 0 <= ny < m + 2 and new_storage[nx][ny] == '0':
                            buffer.append((i,j))
                            break
        for x,y in buffer:
            new_storage[x][y] = '0'
        update()
        
    def crain(request):
        target = request[0]
        nonlocal new_storage
        buffer = []
        for i in range(n + 2):
            for j in range(m + 2):
                if new_storage[i][j] == target:
                    buffer.append((i,j))
        for x,y in buffer:
            new_storage[x][y] = '1'
        update()
        
        
    for r in requests:
        if len(r) == 1:
            fork(r)
        else:
            crain(r)
            
    for row in new_storage:
        print(row)
        
    cnt = 0
    for i in range(1, n + 1):
        for j in range(1, m + 1):
            if new_storage[i][j] != '0' and new_storage[i][j] != '1':
                cnt += 1
    return cnt